Tests/MapboxNavigationTests/VanishingRouteLineTests.swift

```diff
@@ -326,4 +326,72 @@ class VanishingRouteLineTests: TestCase {
+        
+        // When the `NavigationMapView.showsRestrictedAreasOnRoute` is turned on during active navigation with `routeLineTracksTraversal` enabled,
+        // the previous vanishing effect should be kept, and the new restricted areas layer should be added
+        // above the main route line layer but below the arrow stroke layer.
+        navigationMapView.showsRestrictedAreasOnRoute = true
+        do {
+            let layer = try navigationMapView.mapView.mapboxMap.style.layer(withId: layerIdentifier) as! LineLayer
+            let lineGradientString = lineGradientToString(lineGradient: layer.lineGradient)
+            XCTAssertEqual(lineGradientString, expectedExpressionString, "Failed to keep the vanishing effect when showsRestrictedAreasOnRoute turns on.")
+            
+            allLayerIds = navigationMapView.mapView.mapboxMap.style.allLayerIdentifiers.map({ $0.id })
+            guard let indexOfMainRouteLayer = allLayerIds.firstIndex(of: route.identifier(.route(isMainRoute: true))),
+                  let indexOfRestrictedAreas = allLayerIds.firstIndex(of: route.identifier(.restrictedRouteAreaRoute)),
+                  let indexOfArrowStrokeLayer = allLayerIds.firstIndex(of: NavigationMapView.LayerIdentifier.arrowStrokeLayer) else {
+                      XCTFail("Failed to find all the layers")
+                      return
+                  }
+            XCTAssert(indexOfMainRouteLayer < indexOfRestrictedAreas, "Restricted areas route layer should be above main route layer.")
+            XCTAssert(indexOfRestrictedAreas < indexOfArrowStrokeLayer, "Restricted areas route layer should be below arrow stroke layer.")
+        } catch {
+            XCTFail(error.localizedDescription)
+        }
+        
+        // When the `NavigationMapView.showsRestrictedAreasOnRoute` is turned off during active navigation with `routeLineTracksTraversal` enabled,
+        // the previous vanishing effect should be kept, but the restricted areas layer and its source should be removed.
+        navigationMapView.showsRestrictedAreasOnRoute = false
+        do {
+            let layer = try navigationMapView.mapView.mapboxMap.style.layer(withId: layerIdentifier) as! LineLayer
+            let lineGradientString = lineGradientToString(lineGradient: layer.lineGradient)
+            XCTAssertEqual(lineGradientString, expectedExpressionString, "Failed to keep the vanishing effect when showsRestrictedAreasOnRoute turns off.")
+            
+            allLayerIds = navigationMapView.mapView.mapboxMap.style.allLayerIdentifiers.map({ $0.id })
+            XCTAssertFalse(allLayerIds.contains(route.identifier(.restrictedRouteAreaRoute)), "Failed to remove restricted areas route layer.")
+            XCTAssertFalse(allLayerIds.contains(route.identifier(.restrictedRouteAreaSource)), "Failed to remove restricted areas route source.")
```

Comment:
Is it correct check? Should we keep list of layers and sources separately?

Reply:
Thanks for the suggestion, the tests check will be fixed in #3687 .